src/OrchardCore.Modules/OrchardCore.DynamicCache/EventHandlers/DynamicCacheShapeDisplayEvents.cs

```diff
@@ -0,0 +1,102 @@
+
+            // The shape has cache settings and no content yet
+            if (context.ShapeMetadata.IsCached && context.ChildContent == null)
+            {
+                if (debugMode)
+                {
+                    context.ShapeMetadata.Wrappers.Add("CachedShapeWrapper");
+                }
+
+                var cacheContext = context.ShapeMetadata.Cache();
+                _cacheScopeManager.EnterScope(cacheContext);
+
+                var cachedContent = await _dynamicCacheService.GetCachedValueAsync(cacheContext);
+
+                if (cachedContent != null)
+                {
+                    // The contents of this shape was found in the cache.
+                    // Add the cacheContext to _cached so that we don't try to cache the content again in the DisplayedAsync method.
+                    _cached.Add(cacheContext);
+                    context.ChildContent = new HtmlString(cachedContent);
+                }
+            }
+        }
+
+        public async Task DisplayedAsync(ShapeDisplayContext context)
+        {
+            var cacheContext = context.ShapeMetadata.Cache();
+
+            // If the shape is not configured to be cached, continue as usual
+            if (cacheContext == null)
+            {
+                if (context.ChildContent != null)
+                {
+                    string content;
+
+                    using (var sw = new StringWriter())
+                    {
+                        context.ChildContent.WriteTo(sw, HtmlEncoder.Default);
+                        content = sw.ToString();
+                    }
+                    
+                    context.ChildContent = new HtmlString(content);
+                }
+                else
+                {
+                    context.ChildContent = HtmlString.Empty;
+                }
+
+                return;
+            }
+
+            // If we have got this far, then this shape is configured to be cached.
+            // We need to determine whether or not the ChildContent of this shape was retrieved from the cache by the DisplayingAsync method above, as opposed to generated by the View Engine.
+            // ChildContent will be generated by the View Engine if it was not available in the cache when we rendered the shape.
+            // In this instance, we need insert the ChildContent into the cache so that subsequent attempt to render this shape can take advantage of the cached content.
+
+            // If the ChildContent was retrieved form the cache, then the Cache Context will be present in the _cached collection (see the DisplayingAsync method in this class).
+            // So, if the cache context is not present in the _cached collection, we need to insert the ChildContent value into the cache:
+            if (!_cached.Contains(cacheContext) && context.ChildContent != null)
+            {
+                _cacheScopeManager.ExitScope(); // todo: how can we guarantee that this is called, even on failures?
```

Comment:
^^ Any suggestions on how to achieve this?

Reply:
We can also implement `IDisposable` and call dispose on the event instances from the caller. Is it didn't go into `DisplayedAsync` but calls `Dispose` then we know there was an issue and can exit from the scope.

Reply:
Make `IShapeDisplayEvents` implement `IDisposable`? I don't think this would work as these scopes are sensitive to the order that they are opened and closed, so you'd need a new instance of each `IShapeDisplayEvents` for every shape (so that they could be disposed immediately after each shape had been rendered).

I think the first suggestion could work though.

Reply:
ok